--- epan/filesystem.c
+++ epan/filesystem.c
@@ -1035,7 +1035,11 @@
 			 */
 			homedir = g_strdup(pwd->pw_dir);
 		} else
+#ifdef ANDROID
+			homedir = "/cache";
+#else
 			homedir = "/tmp";
+#endif
 	}
 	persconffile_dir = g_strdup_printf("%s" G_DIR_SEPARATOR_S "%s", homedir, PF_DIR);
 #endif
--- wiretap/nettl.c
+++ wiretap/nettl.c
@@ -759,7 +759,11 @@
 	/* Write the file header. */
 	memset(&file_hdr,0,sizeof(file_hdr));
 	memcpy(file_hdr.magic,nettl_magic_hpux10,sizeof(file_hdr.magic));
+#ifdef ANDROID
+	g_strlcpy(file_hdr.file_name,"/cache/wireshark.TRC000",NETTL_FILENAME_SIZE);
+#else
 	g_strlcpy(file_hdr.file_name,"/tmp/wireshark.TRC000",NETTL_FILENAME_SIZE);
+#endif
 	g_strlcpy(file_hdr.tz,"UTC",20);
 	g_strlcpy(file_hdr.host_name,"",9);
 	g_strlcpy(file_hdr.os_vers,"B.11.11",9);
--- wsutil/privileges.c
+++ wsutil/privileges.c
@@ -317,7 +317,9 @@
 	} else {
 		groupname = g_strdup("UNKNOWN");
 	}
+#ifndef ANDROID
 	endgrent();
+#endif
 	return groupname;
 }
 
