--- atk/atkutil.c
+++ atk/atkutil.c
@@ -71,7 +71,7 @@
  * Array of FocusTracker structs
  */
 static GArray *trackers = NULL;
-static guint  index = 0;
+static guint  _index = 0;
 
 typedef struct _FocusTracker FocusTracker;
 
@@ -124,10 +124,10 @@
   {
     FocusTracker item;
 
-    item.index = ++index;
+    item.index = ++_index;
     item.func = focus_tracker;
     trackers = g_array_append_val (trackers, item);
-    return index;
+    return _index;
   }
   else
   {
