--- src/authutil.c
+++ src/authutil.c
@@ -75,7 +75,7 @@
     static char	*buf;
     static int	bsize;
     int	    	size;
-#if defined(WIN32) || defined(__UNIXOS2__)
+#if defined(WIN32) || defined(__UNIXOS2__) || defined(ANDROID)
 #ifndef PATH_MAX
 #define PATH_MAX 512
 #endif
@@ -112,6 +112,11 @@
 	name = dir;
 	if (!name)
 #endif
+#ifdef ANDROID
+        strcpy (dir,"/mnt/sdcard");
+        name = dir;
+        if (!name)
+#endif
 	return (NULL);
     }
 
