--- modules/im/ximcp/imLcIm.c
+++ modules/im/ximcp/imLcIm.c
@@ -59,7 +59,7 @@
 #ifdef COMPOSECACHE
 
 /* include trailing '/' for cache directory, file prefix otherwise */
-#define XIM_GLOBAL_CACHE_DIR "/var/cache/libx11/compose/"
+#define XIM_GLOBAL_CACHE_DIR "/data/cache/libx11/compose/"
 #define XIM_HOME_CACHE_DIR   "/.compose-cache/"
 #define XIM_CACHE_MAGIC      ('X' | 'i'<<8 | 'm'<<16 | 'C'<<24)
 #define XIM_CACHE_VERSION    4
--- src/GetDflt.c
+++ src/GetDflt.c
@@ -127,6 +127,10 @@
 	(void) strncpy(dest, ptr, len-1);
 	dest[len-1] = '\0';
     } else {
+#ifdef ANDROID
+	(void) strncpy(dest, "/mnt/sdcard", len-1);
+	dest[len-1] = '\0';
+#else
 	if ((ptr = getenv("USER")))
 	    pw = _XGetpwnam(ptr,pwparams);
 	else
@@ -136,6 +140,7 @@
 	    dest[len-1] = '\0';
 	} else
 	    *dest = '\0';
+#endif
     }
 #endif
     return dest;
--- src/xlibi18n/lcWrap.c
+++ src/xlibi18n/lcWrap.c
@@ -262,6 +262,9 @@
          * _XlMapOSLocaleName will return the same string or a substring
          * of name, so strlen(name) is okay
          */
+#ifdef ANDROID
+        if (!name) name = "C";
+#endif
         if ((len = strlen(name)) >= sizeof sinamebuf) {
             siname = Xmalloc (len + 1);
             if (siname == NULL)
