--- Xtranssock.c
+++ Xtranssock.c
@@ -207,6 +207,32 @@
 #ifdef UNIXCONN
 
 
+#ifdef ANDROID
+#if defined(X11_t)
+#define UNIX_PATH "/cache/.X11-unix/X"
+#define UNIX_DIR "/cache/.X11-unix"
+#endif /* X11_t */
+#if defined(XIM_t)
+#define UNIX_PATH "/cache/.XIM-unix/XIM"
+#define UNIX_DIR "/cache/.XIM-unix"
+#endif /* XIM_t */
+#if defined(FS_t) || defined(FONT_t)
+#define UNIX_PATH "/cache/.font-unix/fs"
+#define UNIX_DIR "/cache/.font-unix"
+#endif /* FS_t || FONT_t */
+#if defined(ICE_t)
+#define UNIX_PATH "/cache/.ICE-unix/"
+#define UNIX_DIR "/cache/.ICE-unix"
+#endif /* ICE_t */
+#if defined(TEST_t)
+#define UNIX_PATH "/cache/.Test-unix/test"
+#define UNIX_DIR "/cache/.Test-unix"
+#endif
+#if defined(LBXPROXY_t)
+#define UNIX_PATH "/cache/.X11-unix/X"
+#define UNIX_DIR  "/cache/.X11-unix"
+#endif
+#else /* ANDROID */
 #if defined(X11_t)
 #define UNIX_PATH "/tmp/.X11-unix/X"
 #define UNIX_DIR "/tmp/.X11-unix"
@@ -231,6 +257,7 @@
 #define UNIX_PATH "/tmp/.X11-unix/X"
 #define UNIX_DIR  "/tmp/.X11-unix"
 #endif
+#endif /* ANDROID */
 
 
 #endif /* UNIXCONN */
