--- c++/etip.h.in
+++ c++/etip.h.in
@@ -358,7 +358,7 @@
 #endif
 #else
   if (e)
-    cerr << e->message << endl;
+    std::cerr << e->message << std::endl;
   exit(0);
 #endif
 
--- c++/Makefile.in
+++ c++/Makefile.in
@@ -171,7 +171,7 @@
 		-rpath $(INSTALL_PREFIX)$(libdir) \
 		$(LIBTOOL_VERSION) $(NCURSES_MAJOR):$(NCURSES_MINOR) $(LT_UNDEF) $(SHLIB_LIST)
 
-OBJS_DEMO = $(MODEL)/demo$o
+OBJS_DEMO = $(MODEL)/demo$o $(MODEL)/cursesmain$o
 
 $(MODEL)/demo$o : $(srcdir)/demo.cc \
 	$(cursesf_h) $(cursesm_h) $(cursesapp_h)
