--- common/http.h
+++ common/http.h
@@ -25,6 +25,8 @@
 #include "estream.h"
 #endif
 
+#include "../jnlib/strlist.h"
+
 struct uri_tuple_s {
   struct uri_tuple_s *next;
   const char *name;	/* A pointer into name. */
--- config.h.in
+++ config.h.in
@@ -808,6 +808,8 @@
 #define GNUPG_DEFAULT_HOMEDIR "c:/gnupg"
 #elif defined(__VMS)
 #define GNUPG_DEFAULT_HOMEDIR "/SYS\$LOGIN/gnupg"
+#elif defined(ANDROID)
+#define GNUPG_DEFAULT_HOMEDIR "/system/etc/gnupg"
 #else
 #define GNUPG_DEFAULT_HOMEDIR "~/.gnupg"
 #endif
--- tools/gpgkey2ssh.c
+++ tools/gpgkey2ssh.c
@@ -37,6 +37,7 @@
 
 #include "util.h"
 #include "sysutils.h"
+#include "../jnlib/types.h"
 
 
 
